rules/okta_rules/okta_potentially_stolen_session.py

```diff
@@ -0,0 +1,83 @@
+from ast import literal_eval
+from datetime import datetime, timedelta
+from difflib import SequenceMatcher
+
+from panther_base_helpers import deep_get, okta_alert_context
+from panther_oss_helpers import get_string_set, put_string_set, set_key_expiration
+
+PREVIOUS_SESSION = {}
+FUZZ_RATIO_MIN = 0.95
+
+
+def rule(event):
+    # pylint: disable=global-statement
+    # ensure previous session info is avaialable in the alert_context for investigation
+    global PREVIOUS_SESSION
+
+    # Filter only on app access and session start events
+    if (
+        event.get("eventType") == "user.authentication.sso"  # user opened app
+        or event.get("eventType") == "user.session.start"  # user loged in
+    ):
+
+        # lookup if we've previously stored the session cookie
+        PREVIOUS_SESSION = get_string_set(
+            deep_get(event, "authenticationContext", "externalSessionId"),
+        )
+
+        # For unit test mocks we need to eval the string to a set
+        if isinstance(PREVIOUS_SESSION, str):
+            PREVIOUS_SESSION = literal_eval(PREVIOUS_SESSION)
```

Comment:
What is the correct way to do this? I tried that initially but it ended up casting every character as an individual value within the set. I was weary of eval (opted for literal_eval which is supposed to be slightly better) since this condition would only be hit in a test case and not when the rule is running in an instance.

Reply:
working up example w/ `json.loads()` from the mock. Json String of array -> `set(json.loads(stringified_array))`